pandas/tests/indexes/multi/test_setops.py

```diff
@@ -550,12 +550,17 @@ def test_intersection_with_missing_values_on_both_sides(nulls_fixture):
     tm.assert_index_equal(result, expected)
 
 
-def test_union_nan_got_duplicated():
-    # GH#38977
+@pytest.mark.parametrize("sort", [None, False])
+def test_union_nan_got_duplicated(sort):
+    # GH#38977, GH#49010
     mi1 = MultiIndex.from_arrays([[1.0, np.nan], [2, 3]])
     mi2 = MultiIndex.from_arrays([[1.0, np.nan, 3.0], [2, 3, 4]])
-    result = mi1.union(mi2)
-    tm.assert_index_equal(result, mi2)
+    result = mi1.union(mi2, sort=sort)
```

Comment:
Should this also work when the data is EA types with NA?

Reply:
Yes, this also works with EA and pd.NA. I parameterized the test to run both float64(np.nan) and Float64(pd.NA).